### PR TITLE
Make warnings logging work with custom steps API

### LINKED ISSUE #1415: Nowarn should win over warnaserror
Warnings that have been silenced with `nowarn` should not be displayed, regardless of the `warnaserror` settings - even if warnaserror comes last. Currently the `warnaserror` setting will always win regardless of order, which means `nowarn` can't be used to silence warnings if the build has `TreatWarningsAsErrors` set globally.

This is the behavior I see from csc:

```csharp
using System;

namespace con
{
    class Program
    {
        static void Main(string[] args)
        {
            throw new Exception();
            Console.WriteLine("Hello World!");
        }
    }
}
```

```sh
#!/usr/bin/env bash

dotnet "/home/sven/bin/sdk/5.0.100-preview.8.20363.2/Roslyn/bincore/csc.dll" \
/reference:/home/sven/bin/packs/Microsoft.NETCore.App.Ref/5.0.0-preview.8.20361.2/ref/net5.0/System.Console.dll \
/reference:/home/sven/bin/packs/Microsoft.NETCore.App.Ref/5.0.0-preview.8.20361.2/ref/net5.0/System.Runtime.dll \
/out:obj/Debug/net5.0/linux-x64/con.dll /target:exe /nologo \
Program.cs "$@"
```
```
~/con$ ./compile.sh
Program.cs(10,13): warning CS0162: Unreachable code detected
~/con$ ./compile.sh /warnaserror
Program.cs(10,13): error CS0162: Unreachable code detected
~/con$ ./compile.sh /nowarn:CS0162 /warnaserror
~/con$
```

@mateoatr

### PR DESCRIPTION
**Pull request by marek-safar**
As part of the change I also cleaned up uint values usage with only
made the code clumsy.

Also fixes #1415 and possibly other issues

### CODE REVIEW

File: src/linker/Linker/LinkContext.cs [hunk 1]
```diff
@@ -174,9 +174,9 @@ public class LinkContext : IDisposable
 
 		public WarningSuppressionWriter WarningSuppressionWriter { get; private set; }
 
-		public HashSet<uint> NoWarn { get; set; }
+		public HashSet<int> NoWarn { get; set; }
 
-		public Dictionary<uint, bool> WarnAsError { get; set; }
+		public HashSet<int> WarnAsError { get; set; }
```
Comment (sbomer) [inline src/linker/Linker/LinkContext.cs:179]: Note that this also needs to be able to treat individual warnings not as errors to override the global warnaserror setting (https://github.com/mono/linker/issues/1356#issuecomment-658303720).

File: docs/error-codes.md [hunk 2]
```diff
@@ -73,6 +73,10 @@ the error code. For example:
 
 - Invalid value 'value' was used for command-line option '--generate-warning-suppressions'; must be 'cs' or 'xml'.
 
+#### `IL1018`: "Invalid value 'value' was used as warning code
+
+- All warning codes must start with `IL` prefix followed by numeric code.
```
Comment (mateoatr) [inline docs/error-codes.md:78]: Should we mention the valid code range for warnings?

File: src/linker/Linker/MessageContainer.cs [hunk 3]
```diff
@@ -74,14 +64,24 @@ public static MessageContainer CreateErrorMessage (string text, int code, string
...
 		/// warning wave option --warn VERSION. Unversioned warnings are unaffected by this option. </param>
 		/// <returns>New MessageContainer of 'Warning' category</returns>
-		public static MessageContainer CreateWarningMessage (LinkContext context, string text, int code, MessageOrigin origin, string subcategory = MessageSubCategory.None, WarnVersion? version = null)
+		public static MessageContainer CreateWarningMessage (LinkContext context, string text, int code, MessageOrigin origin, WarnVersion version, string subcategory = MessageSubCategory.None)
 		{
-			ValidateWarning (code, version);
+			if (!(code > 2000 && code <= 6000))
+				throw new ArgumentException ($"The provided code '{code}' does not fall into the warning category, which is in the range of 2001 to 6000 (inclusive).");
+
+			if (version >= WarnVersion.ILLink0 && version <= WarnVersion.Latest)
```
Comment (mateoatr) [inline src/linker/Linker/MessageContainer.cs:72]:
```suggestion
			if (!(version >= WarnVersion.ILLink0 && version <= WarnVersion.Latest))
```

File: src/linker/Linker/LinkContext.cs [hunk 4]
```diff
@@ -502,21 +502,6 @@ public void LogMessage (MessageContainer message)
...
-			// Unversioned warnings are not controlled by WarnVersion.
-			// Error messages are guaranteed to only have a version if they were created for a warning due to warnaserror.
-			if ((message.Category == MessageCategory.Warning || message.Category == MessageCategory.Error) &&
-				message.Version > WarnVersion) {
-				// This warning was turned off by --warn.
-				return;
-			}
-
 			if (OutputWarningSuppressions && message.Category == MessageCategory.Warning && message.Origin?.MemberDefinition != null)
 				WarningSuppressionWriter.AddWarning (message.Code.Value, message.Origin?.MemberDefinition);
```
Comment (mateoatr) [inline src/linker/Linker/LinkContext.cs:506]: We can also move this to CreateWarningMessage.

File: docs/error-codes.md (hunk 2 above)
Comment (marek-safar) [reply docs/error-codes.md:78]: Not sure it's worth this, the whole file is about error/warning codes

File: src/linker/Linker/LinkContext.cs (hunk 4 above)
Comment (marek-safar) [reply src/linker/Linker/LinkContext.cs:506]: I think that code is correct here. The warning should go to output file only when it's printed out not if it's constructed.

File: src/linker/Linker/LinkContext.cs (hunk 1 above)
Comment (marek-safar) [reply src/linker/Linker/LinkContext.cs]: Yep, missed that

File: src/linker/Linker/LinkContext.cs [hunk 5]
```diff
@@ -502,21 +502,6 @@ public void LogMessage (MessageContainer message)
...
-			// Note: message.Version is nullable. The comparison is false if it is null.
-			// Unversioned warnings are not controlled by WarnVersion.
-			// Error messages are guaranteed to only have a version if they were created for a warning due to warnaserror.
-			if ((message.Category == MessageCategory.Warning || message.Category == MessageCategory.Error) &&
-				message.Version > WarnVersion) {
-				// This warning was turned off by --warn.
-				return;
-			}
-
 			if (OutputWarningSuppressions && message.Category == MessageCategory.Warning && message.Origin?.MemberDefinition != null)
```
Comment (sbomer) [inline src/linker/Linker/LinkContext.cs:505]: I think this should also output suppressions for for `MessageCategory.WarningAsError`

File: src/linker/Linker/Driver.cs [hunk 6]
```diff
@@ -775,7 +775,7 @@ public int Run (ILogger customLogger = null)
 
 		partial void PreProcessPipeline (Pipeline pipeline);
 
-		private static HashSet<uint> ParseWarnings (string value)
+		private IEnumerable<int> ProcessWarningsCodes (string value)
```
Comment (sbomer) [inline src/linker/Linker/Driver.cs:778]: nit:
```suggestion
		private IEnumerable<int> ProcessWarningCodes (string value)
```

File: src/linker/Linker/MessageContainer.cs [hunk 7]
```diff
@@ -74,14 +64,24 @@ public static MessageContainer CreateErrorMessage (string text, int code, string
 		/// <param name="version">Optional warning version number. Versioned warnings can be controlled with the
 		/// warning wave option --warn VERSION. Unversioned warnings are unaffected by this option. </param>
 		/// <returns>New MessageContainer of 'Warning' category</returns>
-		public static MessageContainer CreateWarningMessage (LinkContext context, string text, int code, MessageOrigin origin, string subcategory = MessageSubCategory.None, WarnVersion? version = null)
+		public static MessageContainer CreateWarningMessage (LinkContext context, string text, int code, MessageOrigin origin, WarnVersion version, string subcategory = MessageSubCategory.None)
 		{
-			ValidateWarning (code, version);
+			if (!(code > 2000 && code <= 6000))
```
Comment (sbomer) [inline src/linker/Linker/MessageContainer.cs:69]: Not sure if you wanted to fix it here, but this doesn't address https://github.com/mono/linker/issues/1384.

**Comment by agocke** [comment]
@marek-safar Can we get this in today? We'd really like this in before the snap EOD

File: src/linker/Linker/MessageContainer.cs (hunk 7 above)
Comment (vitek-karas) [reply src/linker/Linker/MessageContainer.cs:69]: I'm choosing to specifically not address #1384 with this PR - it's related to the intent of these changes, but not really. We need this PR to fix the --nowarn --warnaserror problems, so focusing on that part.

**Comment by sbomer** [comment]
LGTM thanks!

**Pull request merged**